### ISSUE TITLE
should Model.flux() return zero at t<mintime ?

### ISSUE THREAD

**Issue by srodney** [sncosmo/sncosmo#69]
Some of the Whalen et al. Pop III CC SN models include a sharp peak in flux in the very first defined epoch (the shock breakout peak).  These tend to induce a bad Spline2D extrapolation to times preceding that first defined epoch.  For example: 

![popiiisn](https://cloud.githubusercontent.com/assets/1355473/5938594/7fbbe956-a6ca-11e4-93f6-d9039dd6e15f.png)

Physically, the flux in the circled regions should be zero (and magnitudes should be undefined or extremely large) right up to the shock breakout point, which is given by mintime().    

Should the flux() function use a masked array to flag epochs that precede mintime?
Or should it just return zeros?   

Perhaps a new model parameter is needed to define the behavior before mintime?

Should the behavior of flux be adjusted similarly for times after maxtime? (there is less danger of bad extraopolations at late times)

**Comment by kbarbary** [comment]
Yes, `Model.flux()` should indeed return zero before mintime. Unfortunately, there's no option in the `RectBivariateSpline` class to make this happen automatically; for points outside the "bounding box" of the spline, it simply returns the nearest point. For most (all?) other models, this hasn't been an issue because the first phase of the model is at zero flux, so the model does return zero before mintime.

Returning zero flux (rather than a masked array) is the desired behavior: The flux of the model before mintime is not *unknown* (as a masked array would imply). The flux of the model is zero (or close enough to zero that we don't care). Having a well defined flux before mintime make it possible to fit for `t0`: When doing such a fit, the model can shift in time so that the early data points might be either before mintime or after mintime, and we wish to always be able to compare the data points to the model.

The behavior after maxtime is a bit trickier because the models don't usually extend to the point where the flux is close enough to zero that we don't care. Related: #7.

If possible, the easiest thing would be to add a point at zero flux to the Whalen models, slightly before the first epoch. This might induce some crazy spline ringing though. I wish we could add the option directly to `RectBivariateSpline`, but the out of bounds behavior is implemented in the underlying Fortran code; we'd have to PR to scipy and wait for the change to propagate to a release. An alternative is to subclass `RectBivariateSpline` to explicitly check for points before mintime and use this subclass for the Whelan models.

Can you do a quick check to see if adding a point at zero flux just before the first epoch does indeed induce spline ringing? If so, we can go the subclassing route.

**Comment by srodney** [comment]
Yes,  I've checked about adding a zero flux point just before the first epoch in the Whalen models that have a sharp shock breakout peak.   Indeed, the spline fits go crazy, the behavior is highly sensitive to just how closely spaced in time that zero-flux epoch is placed, and it cannot achieves the desired sharpness without disrupting the post-mintime spline behavior. 

Is there a reason that we can't just add a check in the flux (or _flux) functions that explicitly forces flux to zero for all times prior to mintime?   It won't be quite as efficient as building it into the spline function, but it shouldn't be egregiously slow.

**Comment by kbarbary** [comment]
Efficiency was indeed my concern. I've been trying to keep the flux method as fast as possible since it's speed impacts the speed of pretty much everything else (fitting, simulation, etc). I admit that I haven't actually checked to see if it would make a noticeable difference though. It might be fine.

If it does make a big difference, the subclass route is a way to do the check just for models that need it, so one way or another we can solve this.

**Comment by srodney** [comment]
I've done some crude testing, and it looks like adding in a phase check would increase the time for each flux calculation by a factor of about 1.5-2.   I suppose that is a bad enough hit that you don't want to introduce it as the default behavior.   Here is a snippet from the git diff showing my attempt to implement this.  

```
@@ -475,7 +475,11 @@ class TimeSeriesSource(Source):

     def _flux(self, phase, wave):
-        return self._parameters[0] * self._model_flux(phase, wave)
+        phasegrid = np.tile(phase,(len(wave),1)).T
+        return self._parameters[0] * \
+               np.where(phasegrid>=self.minphase(),
+                        self._model_flux(phase,wave),
+                        np.zeros(phasegrid.shape))
```

Maybe there is a more cleverly efficient way to do this (avoiding the array transpose, perhaps?)

**Comment by kbarbary** [comment]
How about this (completely untested code):
```python
def _flux(self, phase, wave):
    f = self._parameters[0] * self._model_flux(phase, wave)
    mask = phase < self.minphase()
    f[mask, :] = 0.
    return f
```

**Comment by srodney** [comment]
Great.  Thats much cleaner and faster.   I tested it, and I'm getting that this slows down the flux computation by no more than ~5% for a TimeSeriesSource.    

Implemented in PR #72 .
